File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-8.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<int []> stack = new Stack<>();
        int maxValue = 0;

        for(int i = 0; i < heights.length; i++){
            if(stack.isEmpty()){
                stack.push(new int []{i, heights[i]});
                continue;
            }
            int index = i;

            while(!stack.isEmpty() && heights[i] < stack.peek()[1]){
                int [] current = stack.pop();
                int distance = (i - current[0]);
                index = current[0];
                maxValue = Math.max(maxValue, distance * current[1]);
            }

            stack.push(new int []{index, heights[i]});
        }

        int maxLength = heights.length;

        while(!stack.isEmpty()){
            int [] current  = stack.pop();
            int distance = maxLength - current[0];
            maxValue = Math.max(maxValue, current[1] * distance);
        }

        return maxValue;
    }
}

File: Data Structures & Algorithms/max-water-container/submission-8.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        left = 0
        right = len(heights) -1

        max_value = 0

        while left < right:

            distance = abs(left - right)
            min_value = min(heights[left], heights[right])

            current_result = distance * min_value

            max_value = max(current_result, max_value)


            if heights[left] <= heights[right]:
                left += 1
            else:
                right -= 1
        
        return max_value
